packages/libs/error-stack/src/ext/stream.rs

```diff
@@ -0,0 +1,361 @@
+    /// [`Report::attach`]: crate::Report::attach
+    fn attach<A>(self, attachment: A) -> StreamWithAttachment<Self, A>
+    where
+        A: Send + Sync + 'static;
+
+    /// Lazily adds a new attachment to the [`Report`] inside the [`Result`] calling
+    /// [`Stream::poll_next`].
+    ///
+    /// Applies [`Report::attach`] to the [`Err`] variant, refer to it for more information.
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::attach`]: crate::Report::attach
+    fn attach_lazy<A, F>(self, attachment: F) -> StreamWithLazyAttachment<Self, F>
+    where
+        A: Send + Sync + 'static,
+        F: Fn() -> A;
+
+    /// Adds a new printable attachment to the [`Report`] inside the [`Result`] when
+    /// calling [`Stream::poll_next`].
+    ///
+    /// Applies [`Report::attach_printable`] to the [`Err`] variant, refer to it for more
+    /// information.
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::attach_printable`]: crate::Report::attach_printable
+    /// [`poll`]: Stream::poll
+    fn attach_printable<A>(self, attachment: A) -> StreamWithPrintableAttachment<Self, A>
+    where
+        A: Display + Debug + Send + Sync + 'static;
+
+    /// Lazily adds a new printable attachment to any [`Report`] in the [`Stream`]
+    /// when calling [`Stream::poll_next`].
+    ///
+    /// Applies [`Report::attach_printable`] to the [`Err`] variant, refer to it for more
+    /// information.
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::attach_printable`]: crate::Report::attach_printable
+    fn attach_printable_lazy<A, F>(
+        self,
+        attachment: F,
+    ) -> StreamWithLazyPrintableAttachment<Self, F>
+    where
+        A: Display + Debug + Send + Sync + 'static,
+        F: Fn() -> A;
+
+    /// Changes the [`Context`] of the [`Report`] inside the [`Result`] when calling
+    /// [`Stream::poll_next`].
+    ///
+    /// Applies [`Report::change_context`] to the [`Err`] variant, see its documentation
+    /// for more information.
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::change_context`]: crate::Report::change_context
+    fn change_context<C>(self, context: C) -> StreamWithContext<Self, C>
+    where
+        C: Context;
+
+    /// Lazily changes the [`Context`] of the [`Report`] inside the [`Result`] when
+    /// calling [`Stream::poll_next`]
+    ///
+    /// Applies [`Report::change_context`] to the [`Err`] variant, see its documntation
+    /// for more information.
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::change_context`]: crate::Report::change_context
+    fn change_context_lazy<C, F>(self, context: F) -> StreamWithLazyContext<Self, F>
+    where
+        C: Context,
+        F: Fn() -> C;
+}
+
+impl<S> StreamReportExt for S
+where
+    S: Stream,
+    S::Item: ResultExt,
+{
+    #[track_caller]
+    fn attach<A>(self, attachment: A) -> StreamWithAttachment<Self, A>
+    where
+        A: Send + Sync + 'static,
```

Comment:
Shouldn't the non-lazy functions require `A`/`C` to be `Clone`? I don't see a reason to call `attach` when the `Stream` implementation cannot be called. Could you also apply this to the `Iterator` extension?